### PR TITLE
PP-891: PBSPro Power Awareness for Cray

### PR DESCRIPTION
**Pull request by ashwathraop**
#### Issue-ID

* **[PP-891](https://pbspro.atlassian.net/browse/PP-891)**

#### Problem description
* Power management functionality for PBS Pro on Cray platform. PBS shall be able to do energy aware scheduling and report energy consumed by the jobs.

#### Solution description
* Extend PBS Professional capabilities to include tracking and accounting of power related metrics. Make PBS Professional able to influence computing resources power consumption and schedule jobs based on power related data. PBS Professional might access power consumption controls and consumption data through platform management tools such as the one provided by SGI.
External design link : [PBSProPowerAwareness](https://pbspro.atlassian.net/wiki/display/PD/PP-735%3A+PBSPro+Power+Awareness)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/lib/Libattr/attr_fn_l.c [hunk 1]
```diff
@@ -206,15 +206,58 @@ set_l(struct attribute *attr, struct attribute *new, enum batch_op op)
...
+ * @retval      0       if ok
+ * @retval     >0       if error
+ *
+ */
+int
+set_l_notneg(struct attribute *attr, struct attribute *new, enum batch_op op)
+{
+	long	val;
+
+	if (!(attr && new && (new->at_flags & ATR_VFLAG_SET)))
```
Comment (vinodchitrali) [inline src/lib/Libattr/attr_fn_l.c]: U can move this check code  action function.  set_l_notneg is not necessary

File: src/lib/Libattr/master_resc_def_all.xml [hunk 2]
```diff
@@ -1008,6 +1009,62 @@
 	<member_at_entlim>PBS_ENTLIM_NOLIMIT</member_at_entlim>
 	<member_at_struct>(struct resource_def *)0</member_at_struct>
    </attributes>
+   <attributes flag = 'SVR'>
```
Comment (vinodchitrali) [inline src/lib/Libattr/master_resc_def_all.xml]: Try to decouple the feature. If we make the resource as custom resource then entire feature can be decoupled from PBSCore  .

File: src/tools/pbs_output.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
```
Comment (vinodchitrali) [inline src/tools/pbs_output.py]: Not sure whether this should be part of PBSPro core package delivery. Personally i feel,  it shouldn't be. Explore the alternative design proposals

File: src/lib/Libattr/attr_fn_l.c (hunk 1 above)
Comment (suresh-thelkar) [reply src/lib/Libattr/attr_fn_l.c]: You can user set_l() as usually and do the rest of the validation that is done here in the corresponding action function.

File: src/lib/Libattr/master_resc_def_all.xml [hunk 4]
```diff
@@ -1008,6 +1009,62 @@
...
+	<member_at_type><SVR>ATR_TYPE_LONG</SVR></member_at_type>
+	<member_at_entlim>PBS_ENTLIM_NOLIMIT</member_at_entlim>
+	<member_at_struct>(struct resource_def *)0</member_at_struct>
+   </attributes>
+   <attributes flag = 'SVR'>
+   /*pcap_accelerator*/
+	<member_name><SVR>"pcap_accelerator"</SVR></member_name>
+	<member_at_decode>decode_l</member_at_decode>
+	<member_at_encode>encode_l</member_at_encode>
+	<member_at_set>set_l_notneg</member_at_set>
```
Comment (suresh-thelkar) [inline src/lib/Libattr/master_resc_def_all.xml]: Use set_l() only wherever required.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+from pbs.v1._pmi_types import BackendError
+from pbs.v1._pmi_utils import _running_excl, _pbs_conf, _get_vnode_names, \
+    _svr_vnode
+
+pbsexec = _pbs_conf("PBS_EXEC")
+if pbsexec is None:
+    raise BackendError("PBS_EXEC not found")
+
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7"))
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7",
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Are we always looking for python2.7 library?

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+
+def launch(id, args):
+    """
+    Run capmc and return the structured output.
+    """
+    import json
+
+    # full path to capmc given by Cray
+    cmd = "/opt/cray/capmc/default/bin/capmc"
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Use os.path.join() to form capmc "cmd".

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    if not os.path.exists(cmd):
+        cmd = "capmc"		# should be in PATH then
+    cmd = cmd + " " + args
+    fail = ""
+
+    pbs.logjobmsg(id, "launch: " + cmd)
+    stdin, stdout, stderr = os.popen3(cmd)
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: collect the output of poopen3() in a list data type.

File: src/lib/Libattr/attr_fn_l.c (hunk 1 above)
Comment (bhroam) [reply src/lib/Libattr/attr_fn_l.c]: What about using the ecl library?  Use the verify_value_zero_or_positive() validation function.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 7 above)
Comment (suresh-thelkar) [reply src/modules/python/pbs/v1/_pmi_cray.py]: I would suggest to use subprocess.Popen() instead of os.popen3() as it offers lot of flexibility so that we need to handle less validations or error cases.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    craynid = "PBScraynid"
+    for vnames in hosts:
+        vnode = _svr_vnode(vnames)
+        try:
+            nidset.add(int(vnode.resources_available[craynid]))
+        except:
+            pass
+    return nidset
+
+
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Describe the function and its arguments as per pep8 guidelines. Please take care of this wherever required.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        energy = 0
+        for node in ret["nodes"]:
+            energy += node["energy_ctr"]
+        pbs.logjobmsg(job.id, "energy usage %dJ" % energy)
+    return energy
+
+
+class Pmi:
+
+    def __init__(self, pyhome=None):
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: If pyhome is not used in this function, remove it as argument.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        pbs.logmsg(pbs.LOG_DEBUG, "Cray: init")
+
+    def _connect(self, endpoint, port):
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Same as the above comment. If endpoint and port are not required please remove it else initialize with default values.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+        e = pbs.event()
+        if e.type == pbs.EXECHOST_PERIODIC:
+            # This function will be called for each job in turn when
+            # running from a periodic hook.  Here we fill in some
+            # global variables just once and use the information
+            # for each job in turn.  Save the result of calling capmc
+            # for all running jobs in the variable ninfo.  Keep a
+            # dictionary with the job id's as keys holding a set
+            # of nid numbers.
+            global ninfo, nidarray
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Make these variables ninfo and nidarray as class level variables which can be access using classname. <variable_name> and acts as global to the class itself.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 10 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: They are kept to match the generic api which is also used for SGI.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 9 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: It is kept to match the generic api which is also used for SGI.

File: src/lib/Libattr/attr_fn_l.c (hunk 1 above)
Comment (ashwathraop) [reply src/lib/Libattr/attr_fn_l.c]: Fixed. Used "zero_or_positive_action" instead along with set_l.

@bhroam Since these attributes get set through hooks I am afraid ecl wouldn't be of much help.

File: src/lib/Libattr/master_resc_def_all.xml (hunk 2 above)
Comment (ashwathraop) [reply src/lib/Libattr/master_resc_def_all.xml]: Decoupling would make the feature very much dependent on Admin. Also we have to consider cray alps which uses few of these attributes to do the reservation.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 5 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Removed it.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 7 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Done.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 11 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Done.

File: src/tools/pbs_output.py (hunk 3 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: This file is a RUR output plugin. Output of this file is used to update resource_used.energy value.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 6 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: I do not see the need to use join() unless we are making a path which could be same on Linux and Windows. In this case it is only Linux and even if we have to support this feature on windows it will be different path. Can you please tell me anything that can cause issues of using path like this?

Comment (mike0042) [reply src/modules/python/pbs/v1/_pmi_cray.py]: The comment from @suresh-thelkar is valid. While what you have will work fine on Linux systems, we use os.path.join() for consistency and portability. Maybe this code will never run on Windows, but if it ever needs porting that will be one less issue to address.

File: src/tools/pbs_output.py (hunk 3 above)
Comment (vinodchitrali) [reply src/tools/pbs_output.py]: Okays. Not sure about this. 

BTW .. did u tried creating RPM  and install ? do u need to update pbs.spec file.

File: src/tools/pbs_output.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ */
+"""
+__doc__ = """
+This script is not intended to be run by users or Administrators.
+It is run by the Cray RUR system as an "output plugin".
+"""
+
+import os
+import sys
+sys.path.append('/opt/cray/rur/default/bin')
```
Comment (vinodchitrali) [inline src/tools/pbs_output.py]: RUR will be  always in this path ?

File: src/tools/pbs_output.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+from rur_plugins import rur_output_args, get_plugin_name, rur_errorlog
+
+
+def outname(jobid):
+    # Create the pathname to write the RUR data for PBS.
+    # By default it will be "/var/spool/PBS/spool/<jobid>.rur"
+    home = "PBS_HOME"
+    dir = "/var/spool/pbs"
```
Comment (vinodchitrali) [inline src/tools/pbs_output.py]: dont hard code the path. If CONF file is not found then exit with error

File: src/tools/pbs_output.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+"""
+__doc__ = """
+This script is not intended to be run by users or Administrators.
+It is run by the Cray RUR system as an "output plugin".
+"""
+
+import os
+import sys
+sys.path.append('/opt/cray/rur/default/bin')
+from rur_plugins import rur_output_args, get_plugin_name, rur_errorlog
```
Comment (vinodchitrali) [inline src/tools/pbs_output.py]: add the error msgs for import fail

File: src/lib/Libattr/attr_fn_l.c [hunk 15]
```diff
@@ -204,17 +204,12 @@ set_l(struct attribute *attr, struct attribute *new, enum batch_op op)
...
- * @param[in] attr - pointer to attribute structure
- * @param[in] with - pointer to attribute structure
- *
- * @return      int
- * @retval      0       if the set of strings in "with" is a subset of "attr"
- * @retval      1       otherwise
+/*
+ * comp_l - compare two attributes of type long
  *
+ *	Returns: +1 if 1st > 2nd
```
Comment (vinodchitrali) [inline src/lib/Libattr/attr_fn_l.c]: is this change related to PR ?

File: src/tools/pbs_output.py (hunk 13 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Thats the default value. We first check for PBS_HOME in os environment if not found we read it from conf file.

File: src/tools/pbs_output.py (hunk 12 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Yes.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 8 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Done.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 6 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Fixed it.

File: src/lib/Libattr/attr_fn_l.c (hunk 15 above)
Comment (ashwathraop) [reply src/lib/Libattr/attr_fn_l.c]: Nope. Restored it.

File: src/tools/pbs_output.py (hunk 14 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Are you suggesting to have try:except block for this import?

**Comment by vinodchitrali** [comment]
Fix codacy issues

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 10 above)
Comment (suresh-thelkar) [reply src/modules/python/pbs/v1/_pmi_cray.py]: @ashwathraop 
I meant to use the same type of initialization of  arguments of class functions i.e. make endpoint and port as default arguments and initialize to None like the way you are doing in  class Power as follows
    def connect(self, endpoint=None, port=None):
        self._check_pmi()
        return self.__pmi._connect(endpoint, port)

The code looks uniform by doing this way.

File: src/tools/pbs_output.py (hunk 12 above)
Comment (vinodchitrali) [reply src/tools/pbs_output.py]: I doubt, i searched in some of the Internet source, it can be installed in any location. Check it

File: src/tools/pbs_output.py (hunk 14 above)
Comment (vinodchitrali) [reply src/tools/pbs_output.py]: yeah; exit on lib unavailability  

Check for  swigify example in pbs_testlib.py line 65;

File: src/tools/pbs_output.py (hunk 13 above)
Comment (vinodchitrali) [reply src/tools/pbs_output.py]: I see that, if PBS_CONF_FILE  is not found then code continues with "/var/spool/pbs". I feel it is wrong.

**Comment by ashwathraop** [comment]
@vinodchitrali I have fixed all the critical codacy issues. Since codacy is not in "required" check for the PR to be merged I left few of them. Let me know what you think of them.

File: src/tools/pbs_output.py (hunk 13 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Its more of an initialization. In the end code does "if os.path.isdir(dir)" and on failure to find this directory it raises an exception.

File: src/tools/pbs_output.py (hunk 14 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Okay.

File: src/tools/pbs_output.py (hunk 12 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: There are few manual steps involved before Admin can start using this script(like updating rur config files). We could document to have the installation directory path to be in system path if it is different from what it is specified in the script.  What do you think?

File: src/lib/Libattr/master_resc_def_all.xml (hunk 4 above)
Comment (ashwathraop) [reply src/lib/Libattr/master_resc_def_all.xml]: Fixed it.

File: src/tools/pbs_output.py (hunk 3 above)
Comment (ashwathraop) [reply src/tools/pbs_output.py]: Fixed it.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 10 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Okay.

**Comment by subhasisb** [comment]
@ashwathraop please rebase for merge

**Comment by ashwathraop** [comment]
@subhasisb Squashed the commits and re-based to the latest. Please take it further.
